### ISSUE TITLE
Simplified acl check for a user

### ISSUE THREAD

**Issue by mstfdkmn** [irods/python-irodsclient#561]
Feature:

It would have been useful to have a relevant method that will return a True/False in access manager to check easily a given user's access level on an object (collection, data object?). Say I have an object and the user A has the "read" access on it and the group A (the user A is member of) has the "own" access. To know whether the user A has the "own" access I need to query the group A. If there are more group based permissions available, then I need to query each of them.

something like:

`session.acls.check_user_acl("bob", "read", "path/to/object")`

I have this solution for a specific need, but I think it might be useful to have a functionality that will work for each access type and for each entity.

  ```
access_rights = []
 with iRODSSession(**zone_environment, password=password) as session:
        obj = session.data_objects.get(object_path)
        for acl in session.acls.get(obj):
            if acl.access_name == "own":
                if acl.user_name == g.irods_session.user.name:
                    access_rights.append(acl.access_name)
            if acl.user_type == "rodsgroup":
                group = session.groups.get(acl.user_name)
                for user in group.members:
                    if user.name == g.irods_session.user.name:
                        access_rights.append(acl.access_name)
    return True if "own" in access_rights else False
```

**Comment by trel** [comment]
A slight tweak/suggestion...

```
session.acls.check(full_logical_path, access_level, user, zone=localZone)
```

- change name to just `check`
- can handle data objects AND collections
- reorder the parameters so that zone is separate (and last) with a default value

---

Should it handle group names too?  hmmm.   not sure the use case...

**Comment by d-w-moore** [comment]
> Should it handle group names too? hmmm. not sure the use case...

The check algorithm is  more trivial for groups.  if `user`  names an iRODS group and `session.acls.get( obj )` fetches any ACLs for this group at `access_level` privilege or higher, then we would return True.

**Comment by d-w-moore** [comment]
> A slight tweak/suggestion...
> 
> ```
> session.acls.check(full_logical_path, access_level, user, zone=localZone)
> ```

Also: I am thinking the `zone` parameter is interpreted  in this call as the zone in which the named `user` lives, (ie `user` refers to **user@zone**), and that if defaulted, in other words our named user is user@zone .  If `zone` takes on the default (probably `'' `or `None`) then I'm assuming the `localZone` we value we should choose should be from the value of `session.zone`.

**Comment by trel** [comment]
Yes, I think `zone` would be the client's `session.zone` (the home zone of the connected user).

**Comment by korydraughn** [comment]
Can a SpecificQuery be put together to accomplish this in one or two calls?

I imagine leaning on the database for this will be significantly more efficient than making multiple iRODS API calls.

**Comment by d-w-moore** [comment]
I imagine we'd have it standard, and baked in to the server - and yes it would probaby be more efficient.

**Comment by korydraughn** [comment]
Correct.

First, we prove a SpecificQuery can be developed to satisfy (or help satisfy) the requirement. If we succeed, then the SpecificQuery is added to the next release of the iRODS server, making it available to all clients.

**Comment by d-w-moore** [comment]
Is it likely to have different text between the different DB flavors, or is this a standard-enough query that one specific query string would do for all the dialects?

**Comment by korydraughn** [comment]
I'm not sure off the top of my head. My guess is there are a few ways to approach the query.

- The query can either solve it completely or partially
- Leveraging more features of the databases could mean better results at the cost of needing database-specific versions of the query
- Perhaps the work behind this becomes an API endpoint, who knows
- There are also database CTEs which might help improve things too

We have to investigate the space.

**Comment by korydraughn** [comment]
Going off of my last comment, I think we need to think about how to best approach this.

Bumping until we have a better grasp on the possible solutions.
